✨ IPv6 support for self-managed clusters (#5603)
kubernetes-sigs/cluster-api-provider-aws: docs/book/src/topics/ipv6-enabled-cluster.md

  ```
</aside>

### CNI IPv6 support

By default, no CNI plugin is installed when provisioning a self-managed cluster. You need to install your own CNI solution that supports IPv6, for example, Calico with VXLAN.

You can find the guides to enable [IPv6](https://docs.tigera.io/calico/latest/networking/ipam/ipv6#ipv6) and [VXLAN](https://docs.tigera.io/calico/latest/networking/configuring/vxlan-ipip) support for Calico on their official documentation. Or you can use a customized Calico manifests [here](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-aws/refs/heads/main/test/e2e/data/cni/calico_ipv6.yaml) for IPv6.

## IPv6 CIDR Allocations

### AWS-assigned IPv6 VPC CIDR

To request AWS to automatically assign an IPv6 CIDR from an AWS defined address pool, use the following setting:

```yaml
spec:
  network:
    vpc:
      ipv6: {}
```

### BYOIPv6 VPC CIDR

To define your own IPv6 address pool and CIDR set the following values:

```yaml
spec:
  network:
    vpc:
      ipv6:
        poolId: pool-id
        cidrBlock: "2009:1234:ff00::/56"
```

There must already be a provisioned pool and a set of IPv6 CIDRs for that.

### BYO IPv6 VPC

> If you have a VPC that is IPv6 enabled (i.e. dual stack VPC) and you would like to use it, please define it in the `AWSCluster` specs:

Reviewer: It might be worth repeating here that while the VPC is dual stack, CAPA will only use IPv6 for the cluster in this configuration.

That does lead me to a question, though - if I have a dual stack VPC, should I be able to run independent IPv4 and IPv6 clusters within it?

Author: > That does lead me to a question, though - if I have a dual stack VPC, should I be able to run independent IPv4 and IPv6 clusters within it?

I haven't tried it yet. But I think, on paper, it should be possible as the IPv6 is "added" to the IPv4 infra. Let me give it a try...

Author: :thought_balloon: I tested with the following setup:

- A dual stack VPC with 3 AZs. There is 1 public and 1 private subnet per AZ.
- Private subnets have NAT64/DNS64 enabled.
- Create 1 IPv6 and 1 IPv4 cluster with the same BYO VPC and subnets (i.e. use the same vpc id and same subnet ids)

**Observations**

The clusters came up healthy and resources have "shared" tags. The destroy also delete resources intended only for that specific cluster.

Checking IPv6 cluster:

```bash
$ kubectl get awscluster -A
NAMESPACE       NAME            CLUSTER         READY   VPC                     BASTION IP   BASTION IPV6
capa-guest-ns   thvo-dev-ipv6   thvo-dev-ipv6   true    vpc-01a97e0f343e69083                
guest-ipv4      thvo-dev-ipv4   thvo-dev-ipv4   true    vpc-01a97e0f343e69083   

$ clusterctl describe cluster thvo-dev-ipv6 
NAME                                                              READY  SEVERITY  REASON  SINCE  MESSAGE                                                                       
Cluster/thvo-dev-ipv6                                             True                     40m                                                                                   
├─ClusterInfrastructure - AWSCluster/thvo-dev-ipv6                True                     48m                                                                                   
├─ControlPlane - KubeadmControlPlane/thvo-dev-ipv6-control-plane  True                     40m                                                                                   
│ └─3 Machines...                                                 True                     42m    See thvo-dev-ipv6-control-plane-c2mg6, thvo-dev-ipv6-control-plane-czpg2, ...  
└─Workers                                                                                                                                                                        
  └─MachineDeployment/thvo-dev-ipv6-md-0                          True                     43m                                                                                   
    └─3 Machines...                                               True                     45m    See thvo-dev-ipv6-md-0-jhvtw-b8g7q, thvo-dev-ipv6-md-0-jhvtw-jt4hk, ...   

$ kubectl --kubeconfig ipv6_kubeconfig get nodes -o=wide
NAME                               STATUS   ROLES           AGE   VERSION   INTERNAL-IP                               EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION   CONTAINER-RUNTIME
i-0127c3ff3594c4a88.ec2.internal   Ready    <none>          47m   v1.32.0   2600:1f18:6b61:d603:8e2c:1a8c:199:41dc    <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
i-0230b5cb184622698.ec2.internal   Ready    control-plane   48m   v1.32.0   2600:1f18:6b61:d603:1de6:6566:8278:16b    <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
i-045f43be8ea2f1914.ec2.internal   Ready    control-plane   42m   v1.32.0   2600:1f18:6b61:d603:c6a6:ade7:92a8:eb4d   <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
i-04bc9b6ce167f10b5.ec2.internal   Ready    control-plane   43m   v1.32.0   2600:1f18:6b61:d604:f7f5:3db3:1fb3:5c3c   <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
i-08b25c5d50207653d.ec2.internal   Ready    <none>          47m   v1.32.0   2600:1f18:6b61:d603:daa6:5cef:e286:7eee   <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
i-0a5c8431f898ab9e9.ec2.internal   Ready    <none>          47m   v1.32.0   2600:1f18:6b61:d603:47fa:4a2a:8a6e:655a   <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
```

Checking Ipv4 cluster:

```bash
$ kubectl get awscluster -A
NAMESPACE       NAME            CLUSTER         READY   VPC                     BASTION IP   BASTION IPV6
capa-guest-ns   thvo-dev-ipv6   thvo-dev-ipv6   true    vpc-01a97e0f343e69083                
guest-ipv4      thvo-dev-ipv4   thvo-dev-ipv4   true    vpc-01a97e0f343e69083   

$ clusterctl describe cluster thvo-dev-ipv4 -n guest-ipv4
NAME                                                              READY  SEVERITY  REASON  SINCE  MESSAGE                                                                       
Cluster/thvo-dev-ipv4                                             True                     6m                                                                                    
├─ClusterInfrastructure - AWSCluster/thvo-dev-ipv4                True                     12m                                                                                   
├─ControlPlane - KubeadmControlPlane/thvo-dev-ipv4-control-plane  True                     6m                                                                                    
│ └─3 Machines...                                                 True                     11m    See thvo-dev-ipv4-control-plane-5cvs6, thvo-dev-ipv4-control-plane-gj996, ...  
└─Workers                                                                                                                                                                        
  └─MachineDeployment/thvo-dev-ipv4-md-0                          True                     8m11s                                                                                 
    └─3 Machines...                                               True                     9m44s  See thvo-dev-ipv4-md-0-9p4d4-hpclt, thvo-dev-ipv4-md-0-9p4d4-lcdwk, ...

$ kubectl --kubeconfig kubeconfig get nodes -o=wide
NAME                           STATUS   ROLES           AGE     VERSION   INTERNAL-IP    EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION   CONTAINER-RUNTIME
ip-10-0-131-64.ec2.internal    Ready    <none>          11m     v1.32.0   10.0.131.64    <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
ip-10-0-136-131.ec2.internal   Ready    <none>          11m     v1.32.0   10.0.136.131   <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
ip-10-0-136-223.ec2.internal   Ready    control-plane   13m     v1.32.0   10.0.136.223   <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
ip-10-0-141-220.ec2.internal   Ready    <none>          11m     v1.32.0   10.0.141.220   <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
ip-10-0-149-43.ec2.internal    Ready    control-plane   8m47s   v1.32.0   10.0.149.43    <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
ip-10-0-154-217.ec2.internal   Ready    control-plane   7m29s   v1.32.0   10.0.154.217   <none>        Ubuntu 24.04.2 LTS   6.8.0-1028-aws   containerd://1.7.20
```

Reviewer: @tthvo Not a specific concern, just trying to think of questions a user may ask.

kind: Cluster
metadata:
  name: "${CLUSTER_NAME}"
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - 192.168.0.0/16
      - fd01::/48
    services:
      cidrBlocks:
      - 172.30.0.0/16
      - fd02::/112
```

## Cloud Controller Manager IPv6 Support for Self-managed Clusters

<aside class="note warning">
<h1>Warning</h1>

The AWS Cloud Controller Manager (CCM) does not currently support dualstack Load Balancers. When creating Services of type LoadBalancer in a dualstack cluster, they will be
assigned addresses from only one IP family.
</aside>

**Node IP addresses**: You need to provide cloud-config to the CCM via a ConfigMap to set the `NodeIPFamilies` to include IPv6. This instructs the CCM to consider IPv6 in the node's network interface. If not, the CCM will only consider node's IPv4. This causes nodes to have only IPv4 and new pods with `hostNetwork: true` will only pick up the node's IPv4 address.

For example, provide the following ConfigMap to `cloud-controller-manager-addon`:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cloud-config
  namespace: kube-system
data:
  cloud-config.conf: |
    [Global]
    NodeIPFamilies=ipv4

>     NodeIPFamilies=ipv6

Reviewer: This is a little confusing - L210-211 says that only one IP family can be used on a load balancer, but this example shows a cloud.conf with both.

We might want to reword L210 to explicitly say that this configures the CCM's to create dedicated LBs for each family.

Author: Thanks, addressed! I separated the mentioning of node IP and Service LB (managed by CCM) into separate sections now.

# Enabling IPv6

## Overview

> CAPA enables you to create IPv6 and dualstack (IPv4 + IPv6) Kubernetes clusters on Amazon Web Services (AWS) on a dualstack network infrastructure.

Reviewer 2: What about this?
```suggestion
CAPA enables you to create single-stack IPv6 and dualstack (IPv4 + IPv6) Kubernetes clusters on Amazon Web Services (AWS) on a dualstack network infrastructure.
```

Author: Thanks, addressed to `IPv6 only`.

If you want to allocate an IPv6 CIDR to the VPC from an existing VPC IPAM pool, define the pool ID and a prefix length as follows:

```yaml
spec:
  network:
    vpc:
      ipv6:
        ipamPool:
          id: ipam-pool-id
          netmaskLength: 56
```

By default, if you omit `netmaskLength`, CAPA will set it to the default `56`.

### Bring-your-own IPv6 VPC

If you have an existing dualstack VPC that you would like to use, you must explicitly provide the IPv6 CIDR block and egress-only internet gateway ID specs:

```yaml
spec:
  network:
    vpc:
      id: vpc-1234567890abcdefg
      cidrBlock: 10.0.0.0/16
      ipv6:
        cidrBlock: "2001:1234:ff00::/56"
        egressOnlyInternetGatewayId: eigw-1234567890abcdefg
```

This has to be done to explicitly express the user intention to use the IPv6 capabilities of the VPC.

## Creating IPv6 EKS-managed Clusters

To quickly deploy an IPv6 EKS cluster, use the [IPv6 EKS cluster template](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-aws/refs/heads/main/templates/cluster-template-eks-ipv6.yaml).

<aside class="note warning">

<h1>Warning</h1>

> EKS currently only supports IPv6-only clusters (not dualstack). You can't define custom Pod CIDRs on EKS with IPv6. EKS automatically assigns an address range from a unique local address range of `fc00::/7`.

Reviewer 2: if I followed correctly the verbiage standard around that feature, it uses "IPv6 only", "dual-stack" (or "dual stack") for the feature references, and `dualstack` when referring to the option/attribute's value. Is that makes sense?
```suggestion
EKS currently only supports IPv6 only clusters (not dual-stack). You can't define custom Pod CIDRs on EKS with IPv6. EKS automatically assigns an address range from a unique local address range of `fc00::/7`.
```

[1] https://docs.aws.amazon.com/eks/latest/best-practices/ipv6.html
[2] https://docs.aws.amazon.com/vpc/latest/userguide/aws-ipv6-support.html

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: "${CLUSTER_NAME}"
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - fd01::/48
    services:
      cidrBlocks:
      - fd02::/112
```

<aside class="note warning">

<h1>Warning</h1>

**DNS64/NAT64**: If you are configuring CAPA to create dualstack private subnets (by default) for an IPv6 cluster and need IPv6-only pods to reach IPv4-only internet services, you must enable [DNS64/NAT64](https://docs.aws.amazon.com/vpc/latest/userguide/nat-gateway-nat64-dns64.html) as CAPA does not do so.

**CoreDNS**: Since CoreDNS pods run on the single-stack IPv6 pod network, they will fail to resolve non-cluster DNS queries via the IPv4 upstream nameserver in `/etc/resolv.conf`.

The workaround is to edit the `coredns` ConfigMap in namespace `kube-system` to use Route53 Resolver IPv6 nameserver `fd00:ec2::253`, by setting `forward . /etc/resolv.conf` part to `forward . fd00:ec2::253 /etc/resolv.conf`.
  ```bash
  kubectl -n kube-system edit cm/coredns
  ```

> **Note**: This CoreDNS workaround is NOT required for dualstack clusters where pods have both IPv4 and IPv6 addresses.

Author: > I see the CoreOS limitation,

Sorry, for this part, I meant CoreDNS to resolve cluster DNS queries. For IPv6-only cluster, internal pod networking has only IPv6 addresses; thus CoreDNS pods cannot reach IPv4 Route53 nameserver (which is the only nameserver defined in host `/etc/resolv.conf` if running on dual-stack subnets). So the user can either tell CAPA to create Ipv6-only subnets (use IPv6 nameserver) or use the workaround. 

> but are we going to support IPv6 only self-managed clusters (without requiring NAT64)? 

Right, that will be nice to have. But as of now, I think not all internet services (e.g. registries, AWS services) are IPv6 compatible, so I leave NAT creation logic intact.

If the users need to get rid of NAT, they can bring their own infrastructure (vpc, subnets). Or later on, we can add an optional to disable NAT completely. WDYT?

# Enabling IPv6

## Overview

> CAPA enables you to create IPv6 only and dualstack (IPv4 + IPv6) Kubernetes clusters on Amazon Web Services (AWS) on a dualstack network infrastructure.

Author: Thanks, I added a `Glossary` section at the top of the doc to make it clearer.

> **Important**: CAPA does not support in-place migration from IPv4 to dualstack or IPv6. You must create a new cluster.

Reviewer 3: ```suggestion
**Important**: CAPA does not support in-place migration on managed clusters from IPv4 to dualstack or IPv6. You must create a new cluster.
```

## Prerequisites

The instance types for control plane and worker machines must support IPv6. To see a list of instance types that support IPv6 in your region, run the following command:

```bash
aws ec2 describe-instance-types \
  --region <region> \
  --filters "Name=network-info.ipv6-supported,Values=true" \
  --query 'InstanceTypes[].InstanceType'
```

If you want to check whether a specific instance type supports IPv6, run the following command (output `true`):

```bash
aws ec2 describe-instance-types \
  --region <region> \
  --instance-types <instance-type> \
  --query 'InstanceTypes[0].NetworkInfo.Ipv6Supported'
```

If you are launching machines in IPv6-only subnets, the instance type must be Nitro-based. To check whether the instance type is Nitro-based, run the following command (output `"nitro"`):

```bash
aws ec2 describe-instance-types \
  --region <region> \
  --instance-types <instance-type> \
  --query 'InstanceTypes[0].Hypervisor'
```

## Enabling IPv6 capabilities

> To instruct CAPA to configure IPv6 capabilities for the network infrastructure, you must explicitly define `spec.network.vpc.ipv6` in either `AWSCluster` or `AWSManagedControlPlane`. See [IPv6 CIDR Allocations](#ipv6-cidr-allocations) for different IPv6 CIDR configuration options.

Reviewer 3: ```suggestion
To instruct CAPA to configure IPv6 capabilities for the cluster's network infrastructure, you must explicitly define `spec.network.vpc.ipv6` in either `AWSCluster` or `AWSManagedControlPlane`. See [IPv6 CIDR Allocations](#ipv6-cidr-allocations) for different IPv6 CIDR configuration options.
```

```yaml
spec:
  network:
    vpc:
      ipv6: {}
```

> **Note:** CAPA, by default, will provision a dualstack infrastructure (i.e. dualstack VPC and subnets). However, your Kubernetes cluster can be configured as either IPv6-only or dualstack depending on your pod/service CIDR configuration.

Reviewer 3: > CAPA, by default, will provision a dualstack infrastructure (i.e. dualstack VPC and subnets).

So this is fixed, non configurable?

> However, your Kubernetes cluster can be configured as either IPv6-only or dualstack depending on your pod/service CIDR configuration

Whereas this can be configured?

Reviewer: I read the 2nd sentence as a clarification - by default, when you specify nothing, CAPA will provision dual stack, but you're able to provide CIDR information to customize the behavior.

|---------------|----------------|----------------------|---------------------|--------|-------|
| IPv4 | IPv4 | ipv4 | ipv4 | ✅ Supported | Traditional IPv4 cluster |
| dualstack | dualstack | dualstack | ipv6 | ✅ Supported | Dualstack LB with IPv6 as primary |
| dualstack | dualstack | dualstack | ipv4 | ✅ Supported | Dualstack LB with IPv4 as primary |
| dualstack | dualstack | ipv4 | ipv4 | ✅ Supported | IPv4-only LB on dualstack infrastructure |
| dualstack | IPv4 | dualstack | ipv4 | ✅ Supported | Dualstack LB with IPv4-only control plane |
| dualstack | IPv4 | ipv4 | ipv4 | ✅ Supported | IPv4-only LB with IPv4-only control plane |
| dualstack | IPv6-only | dualstack | ipv6 | ✅ Supported | Dualstack LB with IPv6-only control plane (NAT64/DNS64 enabled automatically for IPv6-only subnets) |
| dualstack | IPv4 | dualstack | ipv6 | ❌ Invalid | Cannot use IPv6 targets when control plane has no IPv6 |
| IPv6-only | * | * | * | ❌ Invalid | NLB requires IPv4 CIDR on subnets |
| * | * | ipv6 | * | ❌ Invalid | ipv6 is not a valid load balancer IP type |

<aside class="note">

<h1>Note</h1>

> DNS64 is only enabled automatically for IPv6-only private subnets that CAPA manages. If you bring your own subnets, you must enable DNS64 manually for your subnets.

Author: > What enables this automatically? CAPA or AWS itself?

Ahh sorry, I guess it sounds a bit confusing. I reworded this paragraph a bit more...

For subnets that are created by CAPA and are configured to be IPv6-only, they will have `enableDNS64` attribute set to `true`. This must be done explicitly by CAPA AWS does not enable it by default.

**References**
- AWS doc: https://docs.aws.amazon.com/vpc/latest/userguide/nat-gateway-nat64-dns64.html
- Commit: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5603/changes/38167fa34c3fd5f7a2c48242262c76f7f4838e8b

To define your own IPv6 address pool and CIDR set the following values:

```yaml
spec:
  network:
    vpc:
      ipv6:
        poolId: pool-id
        cidrBlock: "2009:1234:ff00::/56"
```

### Bring-your-own IPv6 VPC CIDR via VPC Address Manager (VPC IPAM)

If you want to allocate an IPv6 CIDR to the VPC from an existing VPC IPAM pool, define the pool ID and a prefix length as follows:

```yaml
spec:
  network:
    vpc:
      ipv6:
        ipamPool:
          id: ipam-pool-id
          netmaskLength: 56
```

By default, if you omit `netmaskLength`, CAPA will set it to the default `56`.

### Bring-your-own IPv6 VPC

If you have an existing dualstack VPC that you would like to use, you must explicitly provide the IPv6 CIDR block and egress-only internet gateway ID specs:

```yaml
spec:
  network:
    vpc:
      id: vpc-1234567890abcdefg
      cidrBlock: 10.0.0.0/16
      ipv6:
        cidrBlock: "2001:1234:ff00::/56"
        egressOnlyInternetGatewayId: eigw-1234567890abcdefg
```

This has to be done to explicitly express the user intention to use the IPv6 capabilities of the VPC.

## Creating IPv6 EKS-managed Clusters

> To quickly deploy an IPv6 EKS cluster, use the [IPv6 EKS cluster template](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-aws/refs/heads/main/templates/cluster-template-eks-ipv6.yaml).

Author: Addressed, thanks!

<aside class="note warning">

<h1>Warning</h1>

EKS currently only supports IPv6-only clusters (not dualstack). You can't define custom Pod CIDRs on EKS with IPv6. EKS automatically assigns an address range from a unique local address range of `fc00::/7`.

</aside>

**Notes**: All addons **must** be enabled. A working IPv6 cluster configuration defines `spec.network.vpc.ipv6` and all addons as follows:

```yaml
kind: AWSManagedControlPlane
apiVersion: controlplane.cluster.x-k8s.io/v1beta1
metadata:
  name: "${CLUSTER_NAME}-control-plane"
spec:
  network:
    vpc:
      ipv6: {}
  region: "${AWS_REGION}"
  sshKeyName: "${AWS_SSH_KEY_NAME}"
  version: "${KUBERNETES_VERSION}"
  addons:
    - name: "vpc-cni"
      version: "v1.11.0-eksbuild.1"  # Note: Check for latest compatible version
      # this is important, otherwise environment property update will not work
      conflictResolution: "overwrite"
    - name: "coredns"
      version: "v1.8.7-eksbuild.1"  # Note: Check for latest compatible version
    - name: "kube-proxy"
      version: "v1.22.6-eksbuild.1"  # Note: Check for latest compatible version
```

## Creating IPv6 Only Self-managed Clusters

> To quickly deploy an IPv6 only self-managed cluster, use the [IPv6 cluster template](https://raw.githubusercontent.com/kubernetes-sigs/cluster-api-provider-aws/refs/heads/main/templates/cluster-template-ipv6.yaml).

Author: Addressed, thanks!

When creating a self-managed cluster, you can define the IPv6 Pod and Service CIDR. For example, you can define ULA IPv6 range `fd01::/48` for pod networking and `fd02::/112` for service networking.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: "${CLUSTER_NAME}"
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - fd01::/48
    services:
      cidrBlocks:
      - fd02::/112
```

<aside class="note warning">

<h1>Warning</h1>

**CoreDNS**: Since CoreDNS pods run on the single-stack IPv6 pod network, they will fail to resolve non-cluster DNS queries via the IPv4 upstream nameserver in `/etc/resolv.conf`.

The workaround is to edit the `coredns` ConfigMap in namespace `kube-system` to use Route53 Resolver IPv6 nameserver `fd00:ec2::253`, by setting `forward . /etc/resolv.conf` part to `forward . fd00:ec2::253 /etc/resolv.conf`.
  ```bash
  kubectl -n kube-system edit cm/coredns
  ```

> This CoreDNS workaround is NOT required for dualstack clusters where pods have both IPv4 and IPv6 addresses.

Author: Addressed, thanks!

> **DNS64/NAT64**: If you are creating nodes in dualstack subnets for an IPv6 only cluster and need single-stack IPv6 pods to reach IPv4 internet services, you must manually enable [DNS64](https://docs.aws.amazon.com/vpc/latest/userguide/nat-gateway-nat64-dns64.html) for those subnets as CAPA does not do so. See [Mixing subnets of different IP families](#mixing-subnets-of-different-ip-families) on how to tell CAPA to create IPv6-only subnets.

Author: Thanks! I combined this part into section `Supported Network Configurations` to avoid duplicated mentions of NAT64/DNS64.

</aside>

## Creating Dualstack Self-managed Clusters

> To quickly deploy a dualstack self-managed cluster, you can use:

Author: Addressed, thanks!

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: "${CLUSTER_NAME}"
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - fd01::/48
      - 192.168.0.0/16
    services:
      cidrBlocks:
      - fd02::/112
      - 172.30.0.0/16
```

## Cloud Controller Manager (CCM) Node IP Configurations

You need to provide cloud-config to the CCM via a ConfigMap to set the `NodeIPFamilies` to include IPv6. This instructs the CCM to consider IPv6 in the machine network interface, if any. If not configured, the CCM will only consider the node's IPv4 address. This causes nodes to have only IPv4 and new pods with `hostNetwork: true` will only pick up the node's IPv4 address.

For example, provide the following ConfigMap to `cloud-controller-manager-addon`:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cloud-config
  namespace: kube-system
data:
  cloud-config.conf: |
    [Global]
    NodeIPFamilies=ipv6
    NodeIPFamilies=ipv4
```

And then provide the `cloud-config.conf` to the CCM DaemonSet as follows:

```yaml
spec:
  containers:
  - name: aws-cloud-controller-manager

>     image: registry.k8s.io/provider-aws/cloud-controller-manager:v1.28.3

Reviewer: We don't yet have a release for the CCM support - should this be separated into a later PR?

When creating a self-managed cluster, you can define the IPv6 Pod and Service CIDR. For example, you can define ULA IPv6 range `fd01::/48` for pod networking and `fd02::/112` for service networking.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: "${CLUSTER_NAME}"
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - fd01::/48
    services:
      cidrBlocks:
      - fd02::/112
```

<aside class="note warning">

<h1>Warning</h1>

**CoreDNS Resolution Issue**: In Kubernetes IPv6-only clusters, CoreDNS pods are assigned only IPv6 addresses. When resolving external domain names (e.g. `quay.io`), they forward queries to upstream nameservers listed in `/etc/resolv.conf`. However, these nameservers might be defined with only IPv4 addresses, which CoreDNS pods cannot reach since they only have IPv6 connectivity. This causes DNS resolution failures for non-cluster domains.

**Solution**: Configure CoreDNS to use AWS Route53's IPv6 nameserver at `fd00:ec2::253`. Edit the `coredns` ConfigMap in namespace `kube-system` and change the forwarding configuration from `forward . /etc/resolv.conf` to `forward . fd00:ec2::253 /etc/resolv.conf`:

```bash

> kubectl -n kube-system edit cm/coredns

Author: Ahh whoops, I gotta work on my wording :sweat_smile: Let me address this in a follow-up PR when we update the CCM version + remove the CCM warning note.
